src/theory/strings/array_core_solver.cpp

```diff
@@ -0,0 +1,301 @@
+
+    // note that the term could rewrites to a skolem
+    // get proxy variable for the update term as t
+    Node termProxy = d_termReg.getProxyVariableFor(n);
+    Trace("seq-update") << "- " << termProxy << " = " << n << std::endl;
+    std::vector<Node> exp;
+    d_im.addToExplanation(termProxy, n, exp);
+
+    // optimization: add a short cut t == (seq.update n[0] n[1] n[2]) => t[i] ==
+    // n[2][0]
+    Node left = nm->mkNode(SEQ_NTH, termProxy, n[1]);
+    Node right =
+        nm->mkNode(SEQ_NTH, n[2], nm->mkConst(Rational(0)));  // n[2][0]
+    right = Rewriter::rewrite(right);
+    Node lem = nm->mkNode(EQUAL, left, right);
+    Trace("seq-array-debug") << "enter" << std::endl;
+    sendInference(exp, lem);
+
+    // enumerate possible index
+    for (auto nth : d_index_map)
+    {
+      Node seq = nth.first;
+      if (d_state.areEqual(seq, n) || d_state.areEqual(seq, n[0]))
+      {
+        std::set<Node> indexes = nth.second;
+        for (Node j : indexes)
+        {
+          // optimization: add a short cut for special case (seq.update n[0]
+          // n[1] (seq.unit e))
+          if (n[2].getKind() == SEQ_UNIT)
+          {
+            left = nm->mkNode(DISTINCT, n[1], j);
+            Node nth1 = nm->mkNode(SEQ_NTH, termProxy, j);
+            Node nth2 = nm->mkNode(SEQ_NTH, n[0], j);
+            right = nm->mkNode(EQUAL, nth1, nth2);
+            lem = nm->mkNode(IMPLIES, left, right);
+            sendInference(exp, lem);
+          }
+
+          // normal cases
+          left = nm->mkNode(SEQ_NTH, termProxy, j);
+          Node cond = nm->mkNode(
+              AND,
+              nm->mkNode(LEQ, n[1], j),
+              nm->mkNode(
+                  LT,
+                  j,
+                  nm->mkNode(PLUS, n[1], nm->mkNode(STRING_LENGTH, n[2]))));
+          Node body1 = nm->mkNode(SEQ_NTH, n[2], nm->mkNode(MINUS, j, n[1]));
+          Node body2 = nm->mkNode(SEQ_NTH, n[0], j);
+          right = nm->mkNode(ITE, cond, body1, body2);
+          lem = nm->mkNode(EQUAL, left, right);
+          sendInference(exp, lem);
+        }
+      }
+    }
+  }
+}
+
+void ArrayCoreSolver::check(const std::vector<Node>& nthTerms,
+                            const std::vector<Node>& updateTerms)
+{
+  NodeManager* nm = NodeManager::currentNM();
+
+  Trace("seq-array-debug") << "NTH SIZE: " << nthTerms.size() << std::endl;
+  if (Trace.isOn("seq-array-terms"))
+  {
+    for (const Node& n : nthTerms)
+    {
+      Trace("seq-array-terms") << n << std::endl;
+    }
+  }
+  Trace("seq-array-debug") << "UPDATE SIZE: " << updateTerms.size()
+                           << std::endl;
+  if (Trace.isOn("seq-array-terms"))
+  {
+    for (const Node& n : updateTerms)
+    {
+      Trace("seq-array-terms") << n << std::endl;
+    }
+  }
+  Trace("seq-update") << "SequencesArraySolver::check..." << std::endl;
+  d_writeModel.clear();
+  // AJR: i think d_index_map should be cleared at the beginning of each check?
```

Comment:
Please address @ajreynol 's comment and then delete it.

Reply:
Note that I added the clear on line 190, do confirm that this is correct.

Reply:
Yes, it is correct.